### PR TITLE
[docs/test]: Documentation and test in the list_array.cpp

### PR DESCRIPTION
**Pull request by sagarpandyansit**
#### Issue that resolved
https://github.com/TheAlgorithms/C-Plus-Plus/issues/1373

#### Description of Change
- Documentation added.
- Wrote the test cases for all the methods with edge cases.
- Fixed the bugs in various methods.
- Refined the code.
- Efficiency improved.

#### Checklist

- [x] Added description of change
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1375"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please resolve the [merge conflicts](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1375/conflicts).

File: data_structures/list_array.cpp [hunk 1]
```diff
@@ -1,163 +1,226 @@
...
+    // BinarySearch testing
+    assert(L.search(11) == 1); // search for the existing element
+    assert(L.search(12) == 2);
+    assert(L.search(50) == -1); // search for the non-existing element
+}
+
+// Driver code
+int main() {
+    // Testing
+    test();
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:224]:
```suggestion
    test();  // execute the tests
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Well-done, nice work. 👍

File: data_structures/list_array.cpp [hunk 2]
```diff
@@ -1,163 +1,226 @@
...
+ * We can perform basic CRUD operations as well as other operations like sorting etc.
+ *
+ * ### Algorithm
+ * It implements various method like insert, sort, search etc. efficiently.
+ * You can select the operation and methods will do the rest work for you.
+ * You can insert element, sort them in order, search efficiently, delete values and print the list.
  */
 #include <iostream>
 #include <array>
+#include <cassert>
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:16]:
```suggestion
#include <iostream>   /// for io operations
#include <array>      /// for std::array
#include <cassert>   /// for assert
```

File: data_structures/list_array.cpp [hunk 3]
```diff
@@ -1,163 +1,226 @@
...
+    bool isSorted = false; // indicator whether list is sorted or not
+    /*
+     * search an element in the list using binaraySearch.
+     * @param dataArr - list
+     * @param first - pointer to the first element in the remaining list
+     * @param last - pointer to the last element in the remaining list
+     * @param val - element that will be searched
+     * @return index of element in the list if present else -1
+     */
+    int BinarySearch(const std::array<int, 50>& dataArr, int first, int last, int val) {
```
[the remark this answers is not in the document]
Comment (sagarpandyansit) [reply data_structures/list_array.cpp:33]: @Panquesito7, Thanks for valuable suggestions but, the value should not be unsigned as the array(data) might contain negative values.

File: data_structures/list_array.cpp [hunk 4]
```diff
@@ -1,163 +1,233 @@
...
+ * ### Algorithm
+ * It implements various method like insert, sort, search etc. efficiently.
+ * You can select the operation and methods will do the rest work for you.
+ * You can insert element, sort them in order, search efficiently, delete values and print the list.
  */
-#include <iostream>
-#include <array>
+#include <iostream> // for io operations
+#include <array> // for std::array
+#include <cassert> // for assert
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:16]:
```suggestion
#include <iostream> /// for io operations
#include <array>   /// for std::array
#include <cassert> /// for assert
```

File: data_structures/list_array.cpp [hunk 5]
```diff
@@ -1,163 +1,233 @@
...
+    bool isSorted = false; // indicator whether list is sorted or not
+    /**
+     * @brief Search an element in the list using binaraySearch.
+     * @param dataArr list
+     * @param first pointer to the first element in the remaining list
+     * @param last pointer to the last element in the remaining list
+     * @param val element that will be searched
+     * @return index of element in the list if present else -1
+     */
+    int BinarySearch(const std::array<int, 50>& dataArr, const uint64_t& first, const uint64_t& last, const int& val) {
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:33]: Consider using `uint64_t` rather than `int` (same in other parts of the code).

File: data_structures/list_array.cpp [hunk 6]
```diff
@@ -1,163 +1,233 @@
 /**
  * @file list_array.cpp
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:2]:
```suggestion
 * @file
```

File: data_structures/list_array.cpp [hunk 7]
```diff
@@ -1,163 +1,233 @@
...
- * @todo Add documentation
- * @warning The sorting algorithm is erroneous
+ * @brief Implementation of list_array
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:3]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: data_structures/list_array.cpp [hunk 8]
```diff
@@ -1,163 +1,233 @@
...
+ * You can select the operation and methods will do the rest work for you.
+ * You can insert element, sort them in order, search efficiently, delete values and print the list.
  */
-#include <iostream>
-#include <array>
+#include <iostream> // for io operations
+#include <array> // for std::array
+#include <cassert> // for assert
 
+/**
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:18]:
```suggestion
/**
 * @namespace data_structures
 * @brief Algorithms with data structures
 */
namespace data_structures {
/**
 * @namespace list_array
 * @brief Functions for List Array algorithm
 */
namespace list_array {
/**
```

File: data_structures/list_array.cpp [hunk 9]
```diff
@@ -1,163 +1,233 @@
...
+     */
+    void show() {
+        // Going through each element in the list
+        std::cout << '\n';
         for (int i = 0; i < top; i++) {
-            std::cout << data[i] << "\t";
+            std::cout << data[i] << " "; // print the element
         }
     }
 };
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:182]:
```suggestion
};
}  // namespace list_array
}  // namespace data_structures

```

File: data_structures/list_array.cpp [hunk 10]
```diff
@@ -1,163 +1,233 @@
...
 
-int main() {
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
     list L;
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:189]:
```suggestion
    data_structures::list_array::list L;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/list_array.cpp [hunk 11]
```diff
@@ -1,163 +1,249 @@
...
-    int BinarySearch(const std::array<int, 50>& dataArr, int first, int last, int x) {
-        if (last < first) {
+/**
+ * @namespace data_structures
+ * @brief Algorithms with data structures
+ */
+namespace data_structures {
+/**
+ * @namespace list_array
+ * @brief Functions for List Array algorithm
```
Comment (Panquesito7) [inline data_structures/list_array.cpp:25]:
```suggestion
 * @brief Functions for [Dynamic Array](https://en.wikipedia.org/wiki/Dynamic_array) algorithm
```

**Review by Panquesito7 — APPROVED** [review]
Awesome work, I like how you've improved the code so well.
Code and documentation looks good and very refined; LGTM. 😄 👍

**Review by kvedala — APPROVED** [review]
👍
